### ISSUE TITLE
Focus only on preferred cars

### ISSUE THREAD

**Issue by kareem613** [MerlinCooper/iRacingReplayDirector#28]
This is a feature request, but I also mainly have this implemented already. Just posting the issue to see if you're accepting PRs for this resurrected project. Thanks for bringing it back! I thought I was going to be doing this myself and I really didn't want to get to know the depths of the iracing SDK. 

My implementation for this is a new Rule that gets added to the set of direction rules. It just only focuses on the preferred drivers. Behaves a lot like the random rule but it's limited to just the preferred drivers. Only bug right now is related to handling when no preferred drivers are on the track. 

Once fixed, are you accepting PRs?

**Comment by MerlinCooper** [comment]
Well although the tool - as it is - does it's Job I do see a lot of possiblities for improvement. 

That was one reason why I made contact to Vipoo to get previliges to the original repository.

Since then it was more or less just me contributing to the project and I would be happy if other people participate in developing enhancements, doing bug-fixes etc.. 

Therefore I'm happy to accept PR as long as they work together with already available features and do have a corresponding UI if required.

**Comment by mrzoops** [comment]
This is definitely desired. Wishing the highlights could be strictly focused on selected driver/drivers. I hate seeing a great battle of mine only to switch to some random guy who went off track. This would be huge.

**Comment by MerlinCooper** [comment]
> This is definitely desired. Wishing the highlights could be strictly focused on selected driver/drivers. I hate seeing a great battle of mine only to switch to some random guy who went off track. This would be huge.

Yes Kareem wanted to have a look on that. But not sure how much progress he made with that "enhancement".

**Comment by MerlinCooper** [comment]
@kareem613 @mrzoops 

Did spent some time today to improve the behavior when "Focus on preferred drivers" is checked. So far I have seen the list of preferred drivers is already taken into account when selecting the battles but wasn't for incidents. 

With the changes I made only incidents of preferred drivers should be noted now. 

Would appreciate if you could do some testing and provide feedback before I do merge the changes into the master-branch. 
You can find a zip-file with the binaries [here](https://github.com/MerlinCooper/iRacingReplayDirector/releases/download/untagged-c4d93773283f6f7226c7/Draft-iRacingReplayDirector_20210103.5.zip)

**Comment by mrzoops** [comment]
@MerlinCooper @kareem613 Excited to test this out. Your link to the release is showing 'Not Found'. 

I am sure you are aware but 'Focus On preferred drivers' even for battles doesn't strictly follow this. I usually race with my brother and my goal is to have a highlight video where only my car and his car and focused at any time. As it currently stands it will be showing one of us but then randomly cut away to another random car for a while. Or sometimes it will be focusing on the car right in front of me for a long time. I know its a work in progress just letting you know. 

Once I can download the release I will test this.

**Comment by MerlinCooper** [comment]
Understood. Haven't had a detailed view on the battle-selection. Just have seen that some code was verifying against the preferred driver list - but there might be a bug as well. Will try to dig a little deeper into that code. 

Not sure why the link to the pre-release isn't working for you. Did download the zip-file to google drive. 
https://drive.google.com/file/d/1Z3Bn--vy0-EYJu-7FciEbxOHawdbNtGc/view?usp=sharing

**Comment by mrzoops** [comment]
So I tested this out on a race where I crashed a few laps in, and while it followed me most of the race, it cut from me driving down a straight directly to me in the pit lane after the crash. Did not grab the actual incident.

**Comment by MerlinCooper** [comment]
> 
> 
> So I tested this out on a race where I crashed a few laps in, and while it followed me most of the race, it cut from me driving down a straight directly to me in the pit lane after the crash. Did not grab the actual incident.

Well the change I made is only covering that only incidents of the preferred driver are shown - and no incidents of other cars will be picked for a highlight video. 

As far as I was able to re-construct the implementation Vipoo (Dean) did just picks a random event which is meeting the criteria at the point of time were it's evaluated. This results in ReplayDirector is missing overtakes and crashes etc. as covered by issue #3. Unfortunately that will not be an easy / fast re-work. From my understanding it requires a very different approach how scenes are selected. That's going to be a longer lasting project. 

P.S.: Not sure whether and when I'm willing to spent the time a one man-show.

**Comment by MerlinCooper** [comment]
Found some more code to show other cars then the preferred. Should improve the behavior - but doesn't fix it in all situations. Seems that more code-lines would have to be changes. Would appreciate if you could double-check that set of binaries as well. 

Would merge the pull-request to the master branch (and by that to the next weekly Monday release) after receiving your feedback. Do hope that you can access the binaries using the link below. 

https://github.com/MerlinCooper/iRacingReplayDirector/releases/download/untagged-e38cdf73fc593c7dd25f/Draft-iRacingReplayDirector_20210107.3.zip

**Comment by mrzoops** [comment]
> Do hope that you can access the binaries using the link below.
> 
> https://github.com/MerlinCooper/iRacingReplayDirector/releases/download/untagged-e38cdf73fc593c7dd25f/Draft-iRacingReplayDirector_20210107.3.zip

Unfortunately I get the "Not Found" when clicking the link. 

![image](https://user-images.githubusercontent.com/8879242/103908324-55edce80-50d0-11eb-92ef-fd71fb983c07.png)

**Comment by MerlinCooper** [comment]
Is there a GitHub setting which prevents other people to access "Draft-Releases". 

Found another option to publish a pre-release. Can you please try again whether you can access the Assets now?
https://github.com/MerlinCooper/iRacingReplayDirector/releases/tag/20210107.3

**Comment by mrzoops** [comment]
That works. I can access that. I downloaded .7 and will test it very soon. Thank you!

**Comment by MerlinCooper** [comment]
Please make sure that you use the version I linked. Did work on other issues in parallel and therefore not every compiled (pre-released) version does contain the most recent change with regard to the "Preferred Driver issue".

**Comment by mrzoops** [comment]
Ok understood

**Comment by mrzoops** [comment]
OK that test was much much better. Definitely stayed on my car almost the entire race, except for 2 times when it would cut to other cars just sitting in the pit lane. 

Only other odd thing was that it didn't grab my final incident but cut straight to me sitting in the pit.  Very good progress.

**Comment by MerlinCooper** [comment]
All right. Thanks for supporting. 

Based on being a step forward I will put that into the next release (Monday). 

Would you mind sharing your replay and the video generated. Thus I can try to have a look when exactly things still go wrong. Especially showing cars in the pits while others still on track is something which I would like to have a look at. 

P.S.: Towards the end of the race there might be an issue from the very beginning. I always had the issue that it missed drivers getting the chequered flag.

**Comment by mrzoops** [comment]
OK the link to my video is below. I used Build 20210107.3 as originally discussed. The only issues I had were at race start it focuses only on the leader (I tried setting seconds for race leader at 0, but still didn't fix it), and then followed a few pits by random car (might have been the leader). 
Other than that, just the stutters between clips and overlays not working. Other than that it feels like it would be perfect. 

Please let me know if there is anything else I can test for you. 

https://drive.google.com/file/d/1Ul6mfx37jeQq5dhPdi5OrewZ-IKO4CCG/view?usp=sharing

**Comment by MerlinCooper** [comment]
Thanks. Do assume that you are in the yellow car. Correct?

**Comment by mrzoops** [comment]
I am in the bright  yellow/green PUMA car, yes.
